packages/dds/tree/src/util/breakable.ts

```diff
@@ -0,0 +1,129 @@
+
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+
+/**
+ * An object which can enter a "broken" state where trying to use it is a UsageError.
+ */
+export class Breakable {
+	private brokenBy?: Error;
+
+	public constructor(private readonly name: string) {}
+
+	/**
+	 * Throws if the object is in the broken state.
+	 * @remarks
+	 * Can use {@link throwIfBroken} to apply this to a method.
+	 */
+	public use(): void {
+		if (this.brokenBy !== undefined) {
+			throw new UsageError(
+				`Invalid use of ${this.name} after it was put into an invalid state by another error.\nOriginal Error:\n${this.brokenBy}`,
+			);
+		}
+	}
+
+	/**
+	 * Puts this object into the broken state, and throws an error.
+	 *
+	 * @throws If already broken by a different error, throws a UsageError, otherwise throws `brokenBy`.
+	 */
+	public break(brokenBy: Error): never {
+		// If already broken by this error, let it bubble up without rethrowing a modified version.
+		// This prevents internal errors like asserts getting rethrow different errors when wrapped with multiple catches to ensure brokenBy is set.
+		if (this.brokenBy !== brokenBy) {
+			this.use();
+			this.brokenBy = brokenBy;
+		}
+		throw brokenBy;
+	}
+
+	/**
+	 * {@link Breakable.break}, except tolerates `unknown` to be more easily used by catch blocks.
+	 * @privateRemarks
+	 * If there is a use-case, this should be made public.
+	 */
+	private rethrowCaught(brokenBy: unknown): never {
+		if (brokenBy instanceof Error) {
+			this.break(brokenBy);
+		}
+		this.break(
+			new Error(`Non-error throw breaking ${this.name}. Thrown value: "${brokenBy}"`),
+		);
+	}
+
+	/**
+	 * Runs code which should break the object if it throws.
+	 * @remarks
+	 * This also throws if already broken like {@link Breakable.use}.
+	 * Any exceptions this catches are re-thrown.
+	 * Can use {@link breakingMethod} to apply this to a method.
+	 */
+	public run<TResult>(breaker: () => TResult): TResult {
+		this.use();
+		try {
+			return breaker();
+		} catch (error: unknown) {
+			this.rethrowCaught(error);
+		}
+	}
+}
+
+/**
+ * Marks an object as being able to be in a broken state (unknown/unspecified/broken state due to unhandled exception).
+ * @remarks
+ * See decorators {@link breakingMethod} and {@link throwIfBroken} for ease of use.
+ */
+export interface WithBreakable {
+	readonly breaker: Breakable;
+}
+
+/**
+ * Decorator for methods which should break the object when they throw.
+ * @remarks
+ * This also throws if already broken like {@link throwIfBroken}.
+ * See {@link Breakable.run} for details.
+ *
+ * This should be used on methods which modify data that could result in an unsupported/broken state if an exception is thrown while modifying.
+ * It is ok for breakingMethods to call each-other.
+ * @privateRemarks
+ * Explicitly capturing the full `Target` type is necessary to make this work with generic methods with unknown numbers of type parameters.
+ */
+export function breakingMethod<
```

Comment:
Rather than (or in addition to) having decorators for `breakingMethod` and `throwIfBroken`, I wonder what you think of having a decorator for `WithBreakable` classes such that all instance methods are by default `breakingMethod` and `throwIfBroken`. Then we could have decorators for relaxing that on specific methods.

My motivation for this question is my desire to ensure that, when someone adds a new method to a class, they get the safer semantics by default. Otherwise, I'm concerned that we'll forever be missing "coverage".

Reply:
I can't find any official documentation about TypeScript class decorators. I just have https://www.typescriptlang.org/docs/handbook/decorators.html which is only relevant in that it links https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#decorators . That does link https://2ality.com/2022/10/javascript-decorators.html which predates the typescript support but covers the JS class decorator proposal.

They do apparently exists though, https://blog.logrocket.com/practical-guide-typescript-decorators/ has an example.

I think what you suggest is possible. I'll think about it.

Generally, I'm not too worried about missing coverage since the goal here is not so much to be robust, but to help mitigate cases where customers report errors that happen after a fatal error but not the first fatal error (which in practice does seem to occur). If this code does anything, its already undefined behavior, and there already was a fatal error. That said, for cases where there are not a lot of class methods and they are not perf sensitive a class decorator might be a clean way to do it. Gotta be careful to make it work with subclasses and not double wrap things, but it should be possible by putting runtime marker on the wrapped methods.

Reply:
In either case, I'm happy to merge this in as it's clearly an improvement and doing the class thing may not be straightforward.